README.md

golang mysql pool proxy(MPP)
=======
mysql procy pool by golang, select query and use HTTP return data

<a href="README_cn.md">英文说明最详细，简单的可以看中文说明</a>

Future(coming soon...)
======
* golang connect mysql， connect pool， HTTP get data，use POST for sql query 
* support postgre database 
* support many program language (just use POST query) 
* write, read separate 
* deal SQL inject
* support HTTPS 
* distributed HTTP server
* support log file
 
build & run (linux), windows the same
=====
```cd project_path  ```<br />
```go build  ```<br />
```mv golang_mysql_proxy_pool mpp ```<br />
```./mpp ``` <br /><br />

#### how run as daemon & log file
if you want to run as daemon, you can do like this(for linux):

```nohub  ./mpp > mpp.log 2>&1 & ```

that will output the log info to the file mpp.log, more info just man nohub 


#### how to use ? cache & data & ???
the client for PHP in the folder ./example/mpp_cls.php, the sample code 
for every function of MPP client, please check it

Test
=======
WIN7, go version go1.4.2 windows/386, PHP 5.6, MYSQL5.6, 10000 mysql rows
example folder<br /><br />
dont use MPP php file，  php no_use_mpp.php  -----Use time: 1.2660720348358-------<br />
use MPP php file，  php use_mpp.php   -----Use time: 0.37302088737488-------


[more contact<82049406@qq.com>]











CHANGELOG.md
Change Log
===================

1.06
-------------
- 增加检查insert, delete, update的动作并返回影响行数
- 完善了PHP客户端的功能接口，支持目前MPP的所有功能调用 

1.05
-------------
- 增加设置缓存过期时间
- 增加获取缓存
- 增加HTTPS支持
- 完善了PHP的客户端文件，能支持现有的所有功能 

1.04
-------------
- 增加postgre数据库的支持
- 增加postgre数据库配置文件的支持
- 增加守护进程运行和输出log文件的方法 

1.03
-------------
- 增加缓存功能
- 增加测试用例文件

1.02
-------------
- 增加可配置文件选项
- 增加mysql安全链接密钥定义 
- 完善PHP客户端类程序 

1.01
-------------
- 返回查询数据增加json支持，定义状态码规则和返回格式 
- 优化查询数据容错控制逻辑 

1.00
-------------
- 增加HTTP服务端返回查询数据 
- 增加mysql连接池，提高连接性能 
- 增加性能对比的PHP文件，CURL连接MPP和直接连接mysql 



test.go
package main 

import (
  "fmt"
  "time"
  "strconv"
)

func main() {
  now := time.Now().Unix()
  str := "xxx"
  fmt.Println(now)
  fmt.Println(strconv.Itoa(98) + str)
  fmt.Println(strconv.FormatInt(now, 10) + str)
}

main_test.go
/**
* Copyright 0x029 Inc. 
* License: MIT License
* Author: JJyy<82049406@qq.com>
* This file is .... 
**/
package main

import (
	"fmt"
	"testing"
	// "strconv"
)

func TestCc(t *testing.T) {
  k := "name"
  v := "jimmy"
  ccSet(k, v)
  getV := ccGet(k) 
	fmt.Println("val is ", getV)
}



main.go
/**
* Copyright 0x029 Inc.
* License: MIT License
* Author: JJyy<82049406@qq.com>
* main file of MPP
**/
package main

import (
	"os"
	// "io"
	"bufio"
	"fmt"
	"log"
	"path"
	"strings"
)

const (
	//IP          string = "127.0.0.1"
	IP          string = "0.0.0.0"
	PORT        string = "9090"
	CONFIG_FILE string = "mpp.conf"
	VERSION     string = "0.1"
)


var cc = make(map[string]string) //cache map

func readConfig() (new_config map[string]string) {
	config := make(map[string]string)
	dir, _ := path.Split(os.Args[0])
	os.Chdir(dir)
	// path, _ := os.Getwd()
	// config_file, err := os.Open(path + "/" + CONFIG_FILE) //打开文件
	config_file, err := os.Open(CONFIG_FILE) //打开文件
	defer config_file.Close()
	if err != nil {
		fmt.Println(err)
		fmt.Print("Can not read configuration file. now exit\n")
		os.Exit(0)
	}
	// buff := bufio.NewReader(config_file) //读入缓存
	//读取配置文件
	/**
		for {
			line, err := buff.ReadString('\n') //以'\n'为结束符读入一行
			if err != nil || io.EOF == err {
				break
			}
			rs := []rune(line)
			if string(rs[0:1]) == `#` || len(line) < 3 {
				continue
			}
			str_type := string(rs[0:strings.Index(line, " ")])
			detail := string(rs[strings.Index(line, " ")+1 : len(rs)-1])
	    fmt.Println(str_type)
	    fmt.Println(detail)
			config[str_type] = detail
	    fmt.Println(str_type)
	    fmt.Println(detail)
		}
	  **/

	scanner := bufio.NewScanner(config_file)
	for scanner.Scan() {
		var text = scanner.Text()
		// fmt.Println(text)
		if len(text) != 0 {
			// fmt.Println("xxx")
			// os.Exit(1)
			// }
			if string(text[0:1]) == "#" || len(text) < 3 {
				continue
			}

			str_type := strings.TrimSpace(string(text[0:strings.Index(text, " ")]))
			detail := strings.TrimSpace(string(text[strings.Index(text, " ")+1 : len(text)]))
			// fmt.Println(str_type)
			// fmt.Println(detail)
			config[str_type] = detail
		}
	}

	// fmt.Println(config)
	//再次过滤 (防止没有配置文件)
	return verify(config)
	// return config
}

func readConfigyy() {
	file, err := os.Open(CONFIG_FILE)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		fmt.Println(scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}
}

//验证配置文件
func verify(config map[string]string) (config_bak map[string]string) {
	if len(config["bind"]) < 3 {
		config["bind"] = IP
	}
	if len(config["port"]) < 1 {
		config["port"] = PORT
	}
	// fmt.Println(config)
	return config
}

func checkErr(err error) {
	if err != nil {
		panic(err)
		// log.Fatal(err)
		// fmt.Fprintf(os.Stderr, "Fatal error: %s", err.Error())
		// os.Exit(1)
	}
}

func main() {
	startHttpServ()
}


cache.go
/**
* Copyright 0x029 Inc. 
* License: MIT License
* Author: JJyy<82049406@qq.com>
* operate cache for MPP, KV format
**/
package main

import (
  "fmt"
	// "encoding/json"
  "time"
  "strings"
  "strconv"
)

// var cc = make(map[string]string) //cache map

func ccSet(key string, val string) {
  cc[key] = val
}

func ccGet(key string) string {
  // return cc[key]
  // /**
  ccStr := cc[key]
  // ccSli := strings.Split(ccStr, "@mpp_split@")
  ccSli := strings.Split(ccStr, config["cache_split"])
  fmt.Println(ccSli)
  ccTimeStr := ccSli[1]
  ccTimeSli := strings.Split(ccTimeStr, ":")
  fmt.Println(ccTimeSli)
  // sTime := int(ccTimeSli[0])
  sTime, _ := strconv.ParseInt(ccTimeSli[0], 10, 64)
  setTime, _ := strconv.ParseInt(ccTimeSli[1], 10, 64)
  timeNow := time.Now().Unix()
  if ((timeNow - sTime) > setTime && setTime != 0) {
    // fmt.Println("cache out of time")
    return "cache out of time"
  } else {
    // fmt.Println(ccSli[0])
    return ccSli[0]
  }
  return "none"
  // **/
}

func ccDel(key string) {
  delete(cc, key)
}


/**
* return all cache info
*
*@author: JJyy
*@param: 
*@return: 
*
**/
func ccInfo() string {
	// info, _ := json.Marshal(cc)
  // return string(info)
  var reStr string
  for k, v := range cc {
    reStr += k+" "+v+"<br><br>"
    // fmt.Println(k, v)
  }
  return reStr
}






example/mpp_cls.php
<?php 
/**
* Copyright 0x029 Inc. 
* License: MIT License
* Author: JJyy<82049406@qq.com>
* Mpp client PHP class 
**/

class Mpp {
  public $_mppServ; //mpp server addr
  public $_tblPrefix; //table prefix
  protected static $_uukey = "9f89c3e4bee30185648172d9013525cf";
  
  function __construct($mppServ) {
    $this->_mppServ = $mppServ;
    // echo $this->_mppServ;
  }

  /**
  * set table prefix
  *@param: 
  *@return: 
  *
  **/
  function setTblPrefix($prefix) {
    $this->_tblPrefix = $prefix;
  }

  /**
  * 1,fetch data from mpp by query
  * 2,set cache by query
  *
  *@param: string query, array ccArr('key', 'time') 
  *@return: query result
  *
  **/
  function fetch($query, $ccArr=array()) {
    if (empty($ccArr)) {
    $rq = '&query='.$query;
  } else {
    $rq = '&query='.$query.'&cache=1&cache_key='.$ccArr[0].'&cctime='.$ccArr[1];
  }
    $jsonArr = $this->unifyRequest($rq);
    $jsonArr = $this->formatJson($jsonArr);
    return $jsonArr;
  }
  
  /**
  * get cache by key
  *
  *@param:  string key
  *@return: string 
  *
  **/
  function getCache($key) {
    $rq = '&query=cc_get&cache_key='.$key;
    $jsonArr = $this->unifyRequest($rq);
    $jsonArr = $this->formatJson($jsonArr);
    return $jsonArr;
  }
  
  /**
  * set cache by key
  *
  *@param: string key, string val
  *@return: string
  *
  **/
  // function setCache($key, $val) {
  //   $rq = '&query=cc_get&cache_key='.$key;
  //   $jsonArr = $this->unifyRequest($rq);
  //   $jsonArr = $this->formatJson($jsonArr);
  //   return $jsonArr;
  // }
  
  /**
  * get all cache info
  *
  *@param: 
  *@return: 
  **/
  function getAllCc() {
    $rq = '&query=cc_info';
    $jsonArr = $this->unifyRequest($rq);
    return $jsonArr;
  }
  
  
  /**
  * format json string
  *
  *@param:  json string
  *@return: array 
  **/
  function formatJson($jsonStr) {
    $jsonArr = json_decode($jsonStr, true);
    $jsonArr = $jsonArr['rows'];
    return $jsonArr;
  }
  
  /**
  * unify MPP client request code
  *
  *@param:  string $rq 
  *@return: null 
  **/
  function unifyRequest($rq) {
    $cl = curl_init();
    curl_setopt($cl, CURLOPT_URL, $this->_mppServ);
    // echo "query is: ".$rq."\n\n";
    // curl_setopt($cl,CURLOPT_POSTFIELDS,'uukey='.self::$_uukey.'&query='.$query); 
    curl_setopt($cl,CURLOPT_POSTFIELDS,'uukey='.self::$_uukey.$rq); 
    curl_setopt($cl, CURLOPT_RETURNTRANSFER, 1);
    $str = curl_exec($cl);
    curl_close($cl);
    return $str;
  } 
  
  /**
  * insert, update, delete query, return affect numbers
  *
  *@author: JJyy
  *@param:string query 
  *@return:affect rows 
  **/
  function query($query) {
    $rq = '&query='.$query;
    $jsonArr = $this->unifyRequest($rq);
    $affectNum = $this->formatJson($jsonArr);
    return $affectNum;
  }
  
}

$mpp = new Mpp("http://127.0.0.1:9090/mpp");
//no set cache, just query DB
// $str = $mpp->fetch("select * from users limit 1");

//set  cache
// $str = $mpp->fetch("select * from users limit 1", array('sql_key', '3600'));

//get cache
// $str = $mpp->getCache('sql_key');

//get all cache info
// $str = $mpp->getAllCc();

//update the user info
$sql = "update users set email='0123@111.com' where id=1";
$str = $mpp->query($sql);

print_r($str);

?>


pool.go
/**
* Copyright 0x029 Inc.
* License: MIT License
* Author: JJyy<82049406@qq.com>
* mysql pool setting
**/
package main

import (
	"database/sql"
	"fmt"
	_ "github.com/go-sql-driver/mysql"
	_ "github.com/lib/pq"
	_ "github.com/ziutek/mymysql/godrv"
	"log"
	"net/http"
	// "strings"
	"encoding/json"
	// "os"
	"regexp"
	"runtime"
	"strconv"
	"time"
)

// /**
var config = readConfig()

// fmt.Println(config)
// var max_pool = strconv.ParseInt(string(config["max_pool_size"]), 10, 0)
// **/

var MAX_POOL_SIZE = 10
var MySQLPool chan *sql.DB

func getPool() *sql.DB {
	config := readConfig()
	parallel_num, _ := strconv.ParseInt(config["parallel_num"], 10, 0)
	runtime.GOMAXPROCS(int(parallel_num))

	dbAdapter := config["db_adapter"]
	dbhost := config["dbhost"]
	dbport := config["dbport"]
	dbuser := config["dbuser"]
	dbpwd := config["dbpwd"]
	dbname := config["dbname"]
	// fmt.Println(config)
	// fmt.Println(config["bind"])
	if MySQLPool == nil {
		MySQLPool = make(chan *sql.DB, MAX_POOL_SIZE)
	}

	if len(MySQLPool) == 0 { //if no DB conn, we make the pool
		max_pool, _ := strconv.ParseInt(config["max_pool_size"], 10, 0)
		// fmt.Println(max_pool)
		go func() {
			for i := 0; i < int(max_pool); i++ {
				fmt.Println("crean DB conn....")
				// mysqlc, err := sql.Open("mymysql", "tcp:127.0.0.1:3306*test/root/")

				if dbAdapter == "mysql" {
					//mysql conn
					mysqlc, err := sql.Open("mymysql", "tcp:"+dbhost+":"+dbport+"*"+dbname+"/"+dbuser+"/"+dbpwd)
					if err != nil {
						panic(err)
					}
					putPool(mysqlc)
				} else if dbAdapter == "pgsql" {
					//pgsql conn
					// mysqlc, err := sql.Open("postgres", "user=postgres_user password=password host=172.16.2.29 dbname=my_postgres_db sslmode=disable")
					postgrec, err := sql.Open("postgres", "user="+dbuser+" password="+dbpwd+" host="+dbhost+" port="+dbport+" dbname="+dbname+" sslmode=disable")
					if err != nil {
						panic(err)
					}
					putPool(postgrec)
				}
			}
		}()
	}
	return <-MySQLPool //parallel return DB conn use pool
}

func putPool(conn *sql.DB) {
	// fmt.Println("crean DB conn....")
	if MySQLPool == nil {
		MySQLPool = make(chan *sql.DB, MAX_POOL_SIZE)
	}
	if len(MySQLPool) == MAX_POOL_SIZE {
		conn.Close()
		return
	}
	MySQLPool <- conn
}

func mpp(w http.ResponseWriter, r *http.Request) {
	if r.Method == "POST" {
		/**
		*	the json return back
		* code 0 is for success
		*			 1 is for wrong query
		*			 2 is for wrong get data
		* status success or fail
		* rows query data
		**/
		var back = make(map[string]interface{}) //return json
		back["code"] = 0
		back["status"] = "success"
		back["rows"] = ""

		var errCode = 0
		uukey := r.FormValue("uukey")
		if uukey != config["uukey"] {
			back["code"] = 3
			back["status"] = "fail"
			// log.Fatal("uukey not the same")
			errCode = 1
		}

		cache := r.FormValue("cache")     //cache enable
		ccKey := r.FormValue("cache_key") //cache key
		ccTime := r.FormValue("cctime")   //cache time

		if errCode == 0 { //check secure uukey
			query := r.FormValue("query")
			fmt.Println("query is: ", query)
			// os.Exit(3)
			//get cache info
			if query == "cc_info" {	//get all cache info
				// ccInfo()
				cc_info := ccInfo()
				fmt.Fprintf(w, cc_info)
				return
			} else if config["cache"] == "1" && query == "cc_get" { //get cache by key
				// ccGet(ccKey)
				fmt.Fprintf(w, ccGet(ccKey))
				return
			}

			//get DB data
			dbx := getPool() //get the DB conn from pool when we want to use
			//regular check sql insert/update/delete
			querySelect := 1
			matchedI, _ := regexp.MatchString("insert.*", query)
			matchedD, _ := regexp.MatchString("delete.*", query)
			matchedU, _ := regexp.MatchString("update.*", query)
			// fmt.Println("regular query:	", matched)
			//if not select query
			if matchedI || matchedD || matchedU {
				// fmt.Println("not select query")
				stmt, _ := dbx.Prepare(query)
				res, _ := stmt.Exec()
				// fmt.Println("res:	", res)
				// return res.RowsAffected()
				affect, _ := res.RowsAffected()
				// fmt.Println("affect:	", affect)
				back["rows"] = int(affect)
				// back["rows"] = 1 
				querySelect = 0
				//  return
			} else {

				// var rows interface{}
				// rows, errQuery := dbx.Query(query)
				rows, errQuery := dbx.Query(query)
				// rows, affRows, errQuery := dbx.Query(query)
				// fmt.Println("rows:	", rows)
				// fmt.Println("affRows:	", affRows)
				// fmt.Println("affRows:	", rows.RowsAffected())

				if errQuery != nil {
					back["code"] = 1
					back["status"] = "fail"
					// log.Fatal(err)
					fmt.Println(errQuery)
					errCode = 1
				}

				if errQuery == nil {
					// var email string
					colNames, err := rows.Columns()
					checkErr(err)
					readCols := make([]interface{}, len(colNames))
					writeCols := make([][]byte, len(colNames))
					// writeCols := make([]byte, len(colNames))
					// writeCols := make([]interface{}, len(colNames))
					for i, _ := range writeCols {
						readCols[i] = &writeCols[i]
					}

					result := make([]map[string]interface{}, 0)
					// fmt.Println("len res:", len(result))
					// fmt.Println(result)

					// fmt.Println(colNames)

					for rows.Next() {
						if err := rows.Scan(readCols...); err != nil {
							log.Fatal(err)
							back["code"] = 2
							back["status"] = "fail"
						}

						// fmt.Println(writeCols)
						var tmpStr string
						tmpMap := make(map[string]interface{})

						for i, raw := range writeCols {
							// var tmpStr string
							if raw == nil {
								// result[i] = "\\N"
							} else {
								tmpStr += string(raw)
								tmpMap[colNames[i]] = string(raw)
							}
						}
						result = append(result, tmpMap)
					}

					if err := rows.Err(); err != nil {
						log.Fatal(err)
					}
					// fmt.Println(result)

					// back["rows"] = resStr
					back["rows"] = result
				}
				// defer putPool(dbx)	//put the DB conn into pool after finish used
			}
			defer putPool(dbx) //put the DB conn into pool after finish used
			//just query DB
			jsback, _ := json.Marshal(back)

			//if set cache
			if string(config["cache"]) == "1" && cache == "1" && ccKey != "" && querySelect == 1 {
				startTime := strconv.FormatInt(time.Now().Unix(), 10)
				ccJsback := string(jsback) + string(config["cache_split"]) +
					string(startTime) + ":" + ccTime
				//store cache
				// ccSet(ccKey, string(jsback))
				ccSet(ccKey, string(ccJsback))
			}
			// jsback, _ := json.Marshal(result)
			fmt.Fprintf(w, string(jsback))
			// defer putPool(dbx)
		}
	}
}
